src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/update/actions/RevertPathsSwapAction.java

```diff
@@ -0,0 +1,188 @@
+        HaPathIdsPair oldPathIds = stateMachine.getOldProtectedPathIds();
+
+        if (oldPathIds == null) {
+            return;
+        }
+
+        if (oldPathIds.getForward() != null) {
+            PathId oldForwardPathId = getForwardPathId(oldPathIds);
+            updatePathStatus(stateMachine, oldForwardPathId);
+            updateSubPathStatuses(stateMachine, getForwardSubPathId(oldPathIds));
+
+            log.debug("Swapping back the protected forward path {} with {}",
+                    haFlow.getProtectedForwardPathId(),
+                    oldForwardPathId != null ? oldForwardPathId : NO_PATH);
+
+            haFlow.setProtectedForwardPathId(oldForwardPathId);
+            saveHistory(stateMachine, haFlow.getHaFlowId(), oldForwardPathId);
+        }
+
+        if (oldPathIds.getReverse() != null) {
+            PathId oldReversePathId = getReversePathId(oldPathIds);
+            updatePathStatus(stateMachine, oldReversePathId);
+            updateSubPathStatuses(stateMachine, getReverseSubPathId(oldPathIds));
+
+            log.debug("Swapping back the protected reverse path {} with {}",
+                    haFlow.getProtectedReversePathId(),
+                    oldReversePathId != null ? oldReversePathId : NO_PATH);
+
+            haFlow.setProtectedReversePathId(oldReversePathId);
+            saveHistory(stateMachine, haFlow.getHaFlowId(), oldReversePathId);
+        }
+    }
+
+    private void updatePathStatus(HaFlowUpdateFsm stateMachine, PathId oldPathId) {
+        if (oldPathId != null) {
+            haFlowPathRepository.findById(oldPathId)
+                    .ifPresent(path -> {
+                        if (path.getStatus() != FlowPathStatus.ACTIVE) {
+                            path.setStatus(stateMachine.getOldPathStatus(oldPathId));
+                        }
+                    });
+        }
+    }
+
+    private void updateSubPathStatuses(HaFlowUpdateFsm stateMachine, Collection<PathId> oldSubPathIds) {
+        for (PathId oldSubPathId : oldSubPathIds) {
+            flowPathRepository.findById(oldSubPathId)
+                    .ifPresent(path -> {
+                        if (path.getStatus() != FlowPathStatus.ACTIVE) {
+                            path.setStatus(stateMachine.getOldPathStatus(oldSubPathId));
+                        }
+                    });
+        }
+    }
+
+    private void saveHistory(HaFlowUpdateFsm stateMachine, String haFlowId, PathId pathId) {
+        String pathName = pathId == null ? NO_PATH : format("the path %s", pathId);
+        stateMachine.saveActionToHistory("Ha-flow was reverted to old paths",
+                format("The ha-flow %s was updated with %s", haFlowId, pathName));
+    }
+
+    private PathId getForwardPathId(HaPathIdsPair haPathIdsPair) {
+        return Optional.ofNullable(haPathIdsPair)
+                .map(HaPathIdsPair::getForward)
+                .map(HaFlowPathIds::getHaPathId)
+                .orElse(null);
```

Comment:
a minor thing: you unwrap an empty optional here, but later, on 126, check for null. So, actually the logic of optional is needed later, i.e. it makes sense for return an optional from this method.